File: plc4go/internal/plc4go/modbus/readwrite/model/ModbusPDUReadFifoQueueRequest.go
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
package model

import (
    "encoding/xml"
    "errors"
    "io"
    "plc4x.apache.org/plc4go/v0/internal/plc4go/utils"
)

// The data-structure of this message
type ModbusPDUReadFifoQueueRequest struct {
    FifoPointerAddress uint16
    Parent *ModbusPDU
    IModbusPDUReadFifoQueueRequest
}

// The corresponding interface
type IModbusPDUReadFifoQueueRequest interface {
    LengthInBytes() uint16
    LengthInBits() uint16
    Serialize(io utils.WriteBuffer) error
    xml.Marshaler
}

///////////////////////////////////////////////////////////
// Accessors for discriminator values.
///////////////////////////////////////////////////////////
func (m *ModbusPDUReadFifoQueueRequest) ErrorFlag() bool {
    return false
}

func (m *ModbusPDUReadFifoQueueRequest) FunctionFlag() uint8 {
    return 0x18
}

func (m *ModbusPDUReadFifoQueueRequest) Response() bool {
    return false
}


func (m *ModbusPDUReadFifoQueueRequest) InitializeParent(parent *ModbusPDU) {
}

func NewModbusPDUReadFifoQueueRequest(fifoPointerAddress uint16, ) *ModbusPDU {
    child := &ModbusPDUReadFifoQueueRequest{
        FifoPointerAddress: fifoPointerAddress,
        Parent: NewModbusPDU(),
    }
    child.Parent.Child = child
    return child.Parent
}

func CastModbusPDUReadFifoQueueRequest(structType interface{}) ModbusPDUReadFifoQueueRequest {
    castFunc := func(typ interface{}) ModbusPDUReadFifoQueueRequest {
        if casted, ok := typ.(ModbusPDUReadFifoQueueRequest); ok {
            return casted
        }
        if casted, ok := typ.(*ModbusPDUReadFifoQueueRequest); ok {
            return *casted
        }
        if casted, ok := typ.(ModbusPDU); ok {
            return CastModbusPDUReadFifoQueueRequest(casted.Child)
        }
        if casted, ok := typ.(*ModbusPDU); ok {
            return CastModbusPDUReadFifoQueueRequest(casted.Child)
        }
        return ModbusPDUReadFifoQueueRequest{}
    }
    return castFunc(structType)
}

func (m *ModbusPDUReadFifoQueueRequest) LengthInBits() uint16 {
    lengthInBits := uint16(0)

    // Simple field (fifoPointerAddress)
    lengthInBits += 16

    return lengthInBits
}

func (m *ModbusPDUReadFifoQueueRequest) LengthInBytes() uint16 {
    return m.LengthInBits() / 8
}

func ModbusPDUReadFifoQueueRequestParse(io *utils.ReadBuffer) (*ModbusPDU, error) {

    // Simple Field (fifoPointerAddress)
    fifoPointerAddress, _fifoPointerAddressErr := io.ReadUint16(16)
    if _fifoPointerAddressErr != nil {
        return nil, errors.New("Error parsing 'fifoPointerAddress' field " + _fifoPointerAddressErr.Error())
    }

    // Create a partially initialized instance
    _child := &ModbusPDUReadFifoQueueRequest{
        FifoPointerAddress: fifoPointerAddress,
        Parent: &ModbusPDU{},
    }
    _child.Parent.Child = _child
    return _child.Parent, nil
}

func (m *ModbusPDUReadFifoQueueRequest) Serialize(io utils.WriteBuffer) error {
    ser := func() error {

    // Simple Field (fifoPointerAddress)
    fifoPointerAddress := uint16(m.FifoPointerAddress)
    _fifoPointerAddressErr := io.WriteUint16(16, (fifoPointerAddress))
    if _fifoPointerAddressErr != nil {
        return errors.New("Error serializing 'fifoPointerAddress' field " + _fifoPointerAddressErr.Error())
    }

        return nil
    }
    return m.Parent.SerializeParent(io, m, ser)
}

func (m *ModbusPDUReadFifoQueueRequest) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
    var token xml.Token
    var err error
    token = start
    for {
        switch token.(type) {
        case xml.StartElement:
            tok := token.(xml.StartElement)
            switch tok.Name.Local {
            case "fifoPointerAddress":
                var data uint16
                if err := d.DecodeElement(&data, &tok); err != nil {
                    return err
                }
                m.FifoPointerAddress = data
            }
        }
        token, err = d.Token()
        if err != nil {
            if err == io.EOF {
                return nil
            }
            return err
        }
    }
}

func (m *ModbusPDUReadFifoQueueRequest) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
    if err := e.EncodeElement(m.FifoPointerAddress, xml.StartElement{Name: xml.Name{Local: "fifoPointerAddress"}}); err != nil {
        return err
    }
    return nil
}

File: sandbox/plc4py/plc4py/protocols/modbus/readwrite/ModbusPDUReadFifoQueueRequest.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from dataclasses import dataclass

from plc4py.api.exceptions.exceptions import PlcRuntimeException
from plc4py.api.exceptions.exceptions import SerializationException
from plc4py.api.messages.PlcMessage import PlcMessage
from plc4py.protocols.modbus.readwrite.ModbusPDU import ModbusPDU
from plc4py.protocols.modbus.readwrite.ModbusPDU import ModbusPDUBuilder
from plc4py.spi.generation.ReadBuffer import ReadBuffer
from plc4py.spi.generation.WriteBuffer import WriteBuffer
import math


@dataclass
class ModbusPDUReadFifoQueueRequest(ModbusPDU):
    fifo_pointer_address: int
    # Accessors for discriminator values.
    error_flag: bool = False
    function_flag: int = 0x18
    response: bool = False

    def serialize_modbus_pdu_child(self, write_buffer: WriteBuffer):
        write_buffer.push_context("ModbusPDUReadFifoQueueRequest")

        # Simple Field (fifoPointerAddress)
        write_buffer.write_unsigned_short(
            self.fifo_pointer_address, logical_name="fifoPointerAddress"
        )

        write_buffer.pop_context("ModbusPDUReadFifoQueueRequest")

    def length_in_bytes(self) -> int:
        return int(math.ceil(float(self.length_in_bits() / 8.0)))

    def length_in_bits(self) -> int:
        length_in_bits: int = super().length_in_bits()
        _value: ModbusPDUReadFifoQueueRequest = self

        # Simple field (fifoPointerAddress)
        length_in_bits += 16

        return length_in_bits

    @staticmethod
    def static_parse_builder(read_buffer: ReadBuffer, response: bool):
        read_buffer.push_context("ModbusPDUReadFifoQueueRequest")

        fifo_pointer_address: int = read_buffer.read_unsigned_int(
            logical_name="fifoPointerAddress"
        )

        read_buffer.pop_context("ModbusPDUReadFifoQueueRequest")
        # Create the instance
        return ModbusPDUReadFifoQueueRequestBuilder(fifo_pointer_address)

    def equals(self, o: object) -> bool:
        if self == o:
            return True

        if not isinstance(o, ModbusPDUReadFifoQueueRequest):
            return False

        that: ModbusPDUReadFifoQueueRequest = ModbusPDUReadFifoQueueRequest(o)
        return (
            (self.fifo_pointer_address == that.fifo_pointer_address)
            and super().equals(that)
            and True
        )

    def hash_code(self) -> int:
        return hash(self)

    def __str__(self) -> str:
        pass
        # write_buffer_box_based: WriteBufferBoxBased = WriteBufferBoxBased(True, True)
        # try:
        #    write_buffer_box_based.writeSerializable(self)
        # except SerializationException as e:
        #    raise PlcRuntimeException(e)

        # return "\n" + str(write_buffer_box_based.get_box()) + "\n"


@dataclass
class ModbusPDUReadFifoQueueRequestBuilder(ModbusPDUBuilder):
    fifo_pointer_address: int

    def build(
        self,
    ) -> ModbusPDUReadFifoQueueRequest:
        modbus_pdu_read_fifo_queue_request: ModbusPDUReadFifoQueueRequest = (
            ModbusPDUReadFifoQueueRequest(self.fifo_pointer_address)
        )
        return modbus_pdu_read_fifo_queue_request
